### ISSUE TITLE
Consider making GC.KeepAlive() unnecessary

### ISSUE THREAD

**Issue by Dmitry-Me** [dotnet/runtime#54542]
TL;DR GC.KeepAlive() existence is a very subtle surprise to anyone who doesn't know exactly what it does. And here's the problem which it solves.

So you have a piece of code where you have a reference to an object holding an `IntPtr` to some unmanaged resource. The object class finalizer invalidates the `IntPtr` and a method of that object makes use of `IntPtr` by passing it into some unmanaged code.

You do have an explicit reference to the object in your code. The object's method is running. The CLR can still collect the object right in the middle of object's method running, call the finalizer and make the `IntPtr` invalid thus causing undefined behavior in the unmanaged code.

You have to use `GC.KeepAlive()` to prevent that undesired object collection.

Is that not bad enough? One bonus is it doesn't reproduce under debugger. Another bonus is it doesn't reproduce in debug builds.

So the current behavior is...

1. totally counterintuitive. My code has a reference to the object. The object is reachable from my code. How would GC dare to collect it?
2. causes defects which are reproduced in undeterministic fashion and only outside debugger and only in non-debug builds which actually means "you never see this when debugging, it only appears in production" most of the time.

It would be nice if this undeterministic object collection didn't happen even when `GC.KeepAlive()` is not used. It would be nice to make `GC.KeepAlive()` unnecessary.

How hard would it be to teach CLR to not collect objects which are referenced from currently running block of code? How hard would it be to check that such object does have a finalizer and not consider it for collection until the block it's referenced from ends and it's no longer reachable? That would save a ton of person-hours.

**Comment by huoyaoyuan** [comment]
> 1. The object is reachable from my code.

CLR treats the lifetime ends at last usage, not the code block ends. Code blocks are nothing after compilation.

If GC defers the collection until method end (current debug behavior), performance of large method may regress.

I haven't see any realistic usage of `GC.KeeyAlive` in safe code. Most code use `using`.

**Comment by GrabYourPitchforks** [comment]
I'd strongly recommend using `SafeHandle` instead of `IntPtr` as your interop primitive. Then you don't need to write finalizers and don't need to sprinkle `KeepAlive` throughout your code base. If you're still adamant on dropping down to raw `IntPtr` values in order to milk every last little bit of perf from your application, then you're in an advanced power-user world, and requiring `KeepAlive` for that extremely small percentage of users who absolutely require full control over reachability concerns doesn't seem like the worst thing in the world.

**Comment by Dmitry-Me** [comment]
@GrabYourPitchforks Would `SafeHandle` be protected from being collected unexpectedly early?

**Comment by filipnavara** [comment]
`SafeHandle` can be marshalled all the way into P/Invokes where the runtime ensures it's properly references and that the native code gets the wrapped `IntPtr` from it.

**Comment by cshung** [comment]
> @GrabYourPitchforks Would `SafeHandle` be protected from being collected unexpectedly early?

Well, `SafeHandle` is `IDisposable`, so as long as you follow the pattern of not disposing of it before you are done with it, and then there should be a call to `Dispose()` after the interop call, that would make sure the JIT believes the value is still needed after the use. That works exactly the same way `GC.KeepAlive()` does.

As an added benefit, the `IDisposable` pattern will ensure the native resource are disposed of at a deterministic time when `Dispose()` is called. That makes the program so much easier to reason about. You know exactly when the native memory is freed, opposed to waiting for the finalizer, which may happen at random times.

**Comment by Dmitry-Me** [comment]
@cshung Okay, but what if I don't `Dispose()` it - neither explicitly nor with `using`? Will it be protected for unexpected collection?

**Comment by stephentoub** [comment]
The P/Invoke ensures the SafeHandle remains referenced and will not be available for collection until the P/Invoke returns/
